src/lambda_function/s3_event_config/app.py

```diff
@@ -56,35 +67,83 @@ def add_notification(
         bucket (str): bucket name of the s3 bucket to add the config to
         bucket_key_prefix (str): bucket key prefix for where to look for s3 object notifications
     """
-    s3_client.put_bucket_notification_configuration(
-        Bucket=bucket,
-        NotificationConfiguration={
-            f"{destination_type}Configurations": [
-                {
-                    f"{destination_type}Arn": destination_arn,
-                    "Events": ["s3:ObjectCreated:*"],
-                    "Filter": {
-                        "Key": {
-                            "FilterRules": [
-                                {"Name": "prefix", "Value": f"{bucket_key_prefix}/"}
-                            ]
-                        }
-                    },
-                }
-            ]
-        },
+    existing_bucket_notification_configuration = (
+        s3_client.get_bucket_notification_configuration(Bucket=bucket)
+    )
+    existing_notification_config_for_type = (
+        existing_bucket_notification_configuration.get(
+            f"{destination_type}Configurations", {}
+        )
     )
-    logger.info("Put request completed....")
 
+    new_notification_config = {
+        f"{destination_type}Configurations": [
+            {
+                f"{destination_type}Arn": destination_arn,
+                "Events": ["s3:ObjectCreated:*"],
+                "Filter": {
+                    "Key": {
+                        "FilterRules": [
+                            {"Name": "prefix", "Value": f"{bucket_key_prefix}/"}
+                        ]
+                    }
+                },
+            }
+        ]
+    }
+
+    # If the configuration we want to add isn't there or is different then create a new that contains the new value along with any previous data.
+    if not existing_notification_config_for_type or json.dumps(
+        existing_notification_config_for_type, sort_keys=True
+    ) != json.dumps(
+        new_notification_config[f"{destination_type}Configurations"], sort_keys=True
+    ):
```

Comment:
> The second condition seems to depend on the format of the response object from get_bucket_notification_configuration being the same as the request object of put_bucket_notification_configuration, which could change at some point and isn't always the case with AWS resources.

In the docs they are expected to be the same format:

1. https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3/client/get_bucket_notification_configuration.html
2. https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3/client/put_bucket_notification_configuration.html

> Actually -- since we are overwriting any pre-existing notification configuration for this destination type we don't need to check whether this notification configuration exists or is different. We can skip straight to the step where we create the merged configuration.
Not exactly - We can skip doing an PUT to update the configuration with the same settings by having this check here. This is effectively a 'hasChanged' check.